File: lisca-py/src/lisca_py/domain/schema.py
from __future__ import annotations

SCHEMA_VERSION = 1

ROI_STORE_TEMPLATE = "Pos{pos}_roi.zarr"
BG_STORE_TEMPLATE = "Pos{pos}_bg.zarr"

RAW_ARRAY_PATH_TEMPLATE = "roi/{roi_id}/raw"
BG_ARRAY_PATH_TEMPLATE = "roi/{roi_id}/background"
SEG_ARRAY_PATH_TEMPLATE = "roi/{roi_id}/seg_mask"

INDEX_ROI_IDS_PATH = "index/roi_ids"
INDEX_ROI_BBOXES_PATH = "index/roi_bboxes"
INDEX_ROI_PRESENT_PATH = "index/roi_present"

RAW_AXIS_NAMES = ["t", "c", "z", "y", "x"]
SEG_AXIS_NAMES = ["t", "z", "y", "x"]
BG_AXIS_NAMES = ["t", "c", "z"]


def raw_chunks(shape: tuple[int, int, int, int, int]) -> tuple[int, int, int, int, int]:
    _t, _c, _z, y, x = shape
    return (1, 1, 1, y, x)


def raw_shards(shape: tuple[int, int, int, int, int]) -> tuple[int, int, int, int, int] | None:
    t, c, z, y, x = shape
    shards = (min(64, t), c, z, y, x)
    return None if shards == raw_chunks(shape) else shards


def seg_chunks(shape: tuple[int, int, int, int]) -> tuple[int, int, int, int]:
    _t, z, y, x = shape
    return (1, z, y, x)


def seg_shards(shape: tuple[int, int, int, int]) -> tuple[int, int, int, int] | None:
    t, z, y, x = shape
    shards = (min(64, t), z, y, x)
    return None if shards == seg_chunks(shape) else shards


def mask_chunks(shape: tuple[int, int, int]) -> tuple[int, int, int]:
    _t, y, x = shape
    return (1, y, x)


def mask_shards(shape: tuple[int, int, int]) -> tuple[int, int, int] | None:
    t, y, x = shape
    shards = (min(64, t), y, x)
    return None if shards == mask_chunks(shape) else shards


def bg_chunks(shape: tuple[int, int, int]) -> tuple[int, int, int]:
    _t, _c, _z = shape
    return (1, 1, 1)


def bg_shards(shape: tuple[int, int, int]) -> tuple[int, int, int] | None:
    t, c, z = shape
    shards = (min(256, t), c, z)
    return None if shards == bg_chunks(shape) else shards

File: lisca-rs/src/domain/schema.rs
pub const SCHEMA_VERSION: u64 = 1;

pub fn roi_store_dir(pos: u32) -> String {
    format!("Pos{}_roi.zarr", pos)
}

pub fn bg_store_dir(pos: u32) -> String {
    format!("Pos{}_bg.zarr", pos)
}

pub fn raw_array_path(roi_id: &str) -> String {
    format!("/roi/{}/raw", roi_id)
}

pub fn bg_array_path(roi_id: &str) -> String {
    format!("/roi/{}/background", roi_id)
}

pub fn roi_ids_path() -> &'static str {
    "/index/roi_ids"
}

pub fn roi_bboxes_path() -> &'static str {
    "/index/roi_bboxes"
}

pub fn roi_present_path() -> &'static str {
    "/index/roi_present"
}

pub fn raw_chunks(shape: &[u64]) -> Vec<u64> {
    vec![1, 1, 1, shape[3], shape[4]]
}

pub fn raw_shards(shape: &[u64]) -> Option<Vec<u64>> {
    let shards = vec![shape[0].min(64), shape[1], shape[2], shape[3], shape[4]];
    (shards != raw_chunks(shape)).then_some(shards)
}

pub fn bg_chunks(_shape: &[u64]) -> Vec<u64> {
    vec![1, 1, 1]
}

pub fn bg_shards(shape: &[u64]) -> Option<Vec<u64>> {
    let shards = vec![shape[0].min(256), shape[1], shape[2]];
    (shards != bg_chunks(shape)).then_some(shards)
}

pub fn mask_chunks(shape: &[u64]) -> Vec<u64> {
    vec![1, shape[1], shape[2]]
}

pub fn mask_shards(shape: &[u64]) -> Option<Vec<u64>> {
    let shards = vec![shape[0].min(64), shape[1], shape[2]];
    (shards != mask_chunks(shape)).then_some(shards)
}
